### ISSUE TITLE
Allow collecting handles in a rust Set

### ISSUE THREAD

**Issue by idan-at** [neon-bindings/neon#742]
Hey,

I'm trying to write a native package that needs to "remember" which objects were already seen.

For that purpose, I've tried to use `HashSet` or `BTreeSet`, but `Handle<JsValue>` doesn't implement `Hash`, `Eq` (needed for `HashSet`) or `Ord` (needed for `BTreeSet`).

I found out `neo_runtime` implements `neon_runtime::mem::same_handle`, so I guess it's possible to expose this capability under the hash/eq traits (implementing `Ord`, on the other hand, sounds less trivial).

What do you think? Is that possible?

Basically, this will allow checking whether a given object is circular (which ATM can cause a segmentation fault, due to infinite loops)

Thanks.

**Comment by kjvalencik** [comment]
@idan-at The best I can tell, this isn't possible with N-API. Let me know if you see something in the docs that might be useful: https://nodejs.org/api/n-api.html

The issue is that there's no uniqueness guarantees for handles. You can have two different handles that point to the same object in the JavaScript VM.

The closest is [`napi_strict_equals`](https://nodejs.org/api/n-api.html#n_api_napi_strict_equals) and is available in Neon as [`Handle::strict_equals`](https://docs.rs/neon/0.8.2-napi/neon/handle/struct.Handle.html#method.strict_equals).

I think the best that could be done is collect a set of already seen handles in a `Vec` and iterate across it with each item. However, this is `O(n^2)`. For larger datasets, it might before better to use a JavaScript [`Set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set).

**Comment by idan-at** [comment]
Thanks for the quick reply.

I see [Handle::strict_equals](https://docs.rs/neon/0.8.2-napi/neon/handle/struct.Handle.html#method.strict_equals) works as expected, so thanks for that.

However, as you mentioned, using a vector and iterating over it is inefficient on large collections of objects.

Can you further explain what do you mean by using a Javascript Set? Is it possible to use a Set from a neon library? (In other words, Is it possible using a JS Set when running in Rust?)

Also - do you see any way that we can use the Rust `HashSet` instead? (We can impl `PartialEq` trait by using the `strict_equals` method, but not sure how we can implement the `Hash` trait).

Thanks again for the reply and solution suggestions.

**Comment by kjvalencik** [comment]
I don't think it's possible to use a Rust `HashSet` because it's not possible to get a unique pointer to a value; at least I don't know of one. There can be multiple handles to the same value and each will have a different pointer.

There isn't Neon bindings to `Set` (mostly because there aren't N-API bindings) which makes it a bit awkward, but it can still be done by grabbing a copy from the global object. Quick (untested) example:

```rust
let global = cx.global();
let setConstructor = global.get(&mut cx, "Set")?.downcast_or_throw::<JsFunction, _>(&mut cx)?;

let args: Vec<Handle<JsValue>> = vec![];
let set = setConstructor.construct(&mut cx, args);

let add = set.get(&mut cx, "add")?.downcast_or_throw::<JsFunction, _>(&mut cx)?;
let has = set.get(&mut cx, "has")?.downcast_or_throw::<JsFunction, _>(&mut cx)?;

// Usage
let o = cx.object()?;

add.call(&mut cx, set, vec![o])?;

let has_o = has.call(&mut cx, set, vec![o])?
    .downcast_or_throw::<JsBoolean, _>(&mut cx)?
    .value(&mut cx);
    
println!("Has O: {}", has_o);
```

**Comment by idan-at** [comment]
Thanks for the code snippet. with some (minor) changes, I got it compiled and running. and It actually works, which is awesome (will attach the code at the end, If anyone is interested).

However, fun fact, seems like its slower than the vector implementation (I guess calling JS makes it slower). I've tested on objects with 1000 keys, where the times are almost identical, to 10000000, where its almost a 3 seconds different 😱 

Thanks for the help, I guess I'll stick to the `O(n^2)` vector implementation.

I'd like to ask another question if you don't mind (If you want me to open another issue just LMK).
I've found that the native implementation of retrieve an object's keys is much slower than the JS one:

On the benchmark, I've created an object with X keys (values are `true`), and used `Object.keys` on my JS code, and `get_own_property_names` on my Rust code.

- **10000** keys:
`Object.keys`: 0.3827033042907715
`get_own_property_names`: 2
- **100000** keys:
`Object.keys`: 13.33891487121582
`get_own_property_names`: 32
- **1000000** keys:
`Object.keys`: 206.0333013534546
`get_own_property_names`: 415
- **10000000** keys:
`Object.keys`: 2382.6117510795593
`get_own_property_names`: 8959


Any idea if `get_own_property_names` can be optimized?

Thanks!








JsWeakCode (no error handling, not well optimized, but works ootb)

```rust
struct JsWeakSet<'a> {
    weak_set_instance_handle: Handle<'a, JsObject>,
    add_fn_handle: Handle<'a, JsFunction>,
    has_fn_handle: Handle<'a, JsFunction>,
}

impl<'a> JsWeakSet<'a> {
    pub fn new(cx: &mut FunctionContext<'a>) -> JsWeakSet<'a> {
        let weak_set_constructor_handle = cx
            .global()
            .get(cx, "WeakSet")
            .unwrap()
            .downcast::<JsFunction, _>(cx)
            .unwrap();
        let args: Vec<Handle<JsValue>> = vec![];

        let weak_set_instance_handle = weak_set_constructor_handle.construct(cx, args).unwrap();

        let add_fn_handle = weak_set_instance_handle
            .get(cx, "add")
            .unwrap()
            .downcast::<JsFunction, _>(cx)
            .unwrap();
        let has_fn_handle = weak_set_instance_handle
            .get(cx, "has")
            .unwrap()
            .downcast::<JsFunction, _>(cx)
            .unwrap();

        JsWeakSet {
            weak_set_instance_handle,
            add_fn_handle,
            has_fn_handle,
        }
    }

    pub fn add(&self, cx: &mut FunctionContext<'a>, object: Handle<'a, JsValue>) {
        let args: Vec<Handle<JsValue>> = vec![object];

        self.add_fn_handle
            .call(cx, self.weak_set_instance_handle, args)
            .unwrap();
    }

    pub fn has(&self, cx: &mut FunctionContext<'a>, object: Handle<'a, JsValue>) -> bool {
        let args: Vec<Handle<JsValue>> = vec![object];

        let result = self
            .has_fn_handle
            .call(cx, self.weak_set_instance_handle, args)
            .unwrap();
        result.downcast::<JsBoolean, _>(cx).unwrap().value(cx)
    }
}
```

**Comment by kjvalencik** [comment]
That makes sense. The overhead of making calls across the FFI boundary is incredibly high. There's not much we can do to improve this. Usually the only option is to rewrite code to avoid it. For example, serializing data as `JSON` first.

**Comment by idan-at** [comment]
thanks. have you got a chance to take a look at my other question? (of why `get_own_property_names` is slower than the JS implementation)

LMK if you want me to open a dedicated issue for that

**Comment by kjvalencik** [comment]
It's the same reason. Anything that calls into Js from a native module will be slower than calling it from Js. There's a large fixed overhead, data marshalling and also JIT optimizations are unavailable.

For performance critical native modules it's important to limit the interactions with the VM or they will be a bottleneck.

This isn't something unique to Neon, it applies to all native modules, including C and C++. It's one reason most node database drivers have replaced their native code with highly optimized Js. Even if the protocol is faster in native code, it's not a significant enough of work to overcome the overhead of many fields crossing the boundary.

**Comment by idan-at** [comment]
Wasn't aware of that, I thought calling `N-API` should be faster / at least the same as calling the same method in JS.

Thanks for the explanation!
